library/VTKExtensions/Rendering/vtkF3DRendererWithColoring.cxx

```diff
@@ -259,114 +318,172 @@ int vtkF3DRendererWithColoring::GetColoringComponent()
 //----------------------------------------------------------------------------
 void vtkF3DRendererWithColoring::UpdateColoringActors()
 {
+  assert(this->Importer != nullptr);
+
+  bool importerChanged = this->Importer->GetMTime() >= this->ColoringUpdateTime;
+
   // Early return if nothing changed
-  if (this->ColoringUpdateTime >= this->ColoringTimeStamp.GetMTime())
+  if (!importerChanged && this->ColoringUpdateTime >= this->ColoringTimeStamp.GetMTime())
   {
     return;
   }
 
+  if (importerChanged)
+  {
+    this->ColorTransferFunctionConfigured = false;
+    this->PolyDataMapperConfigured = false;
+    this->PointGaussianMapperConfigured = false;
+    this->VolumeConfigured = false;
+    this->ScalarBarActorConfigured = false;
+  }
+
   this->ColoringUpdateTime = this->ColoringTimeStamp.GetMTime();
 
-  bool volumeVisibility = !this->UseRaytracing && this->UseVolume;
-  if (this->ArrayForColoring || volumeVisibility)
+  // Recover coloring information
+  vtkF3DGenericImporter::ColoringInfo info;
+  bool hasColoring =
+    this->Importer->GetInfoForColoring(this->UseCellColoring, this->ArrayIndexForColoring, info);
+
+  bool volumeVisible = !this->UseRaytracing && this->UseVolume;
+  if (!hasColoring && volumeVisible)
   {
     // When showing volume, always try to find an array to color with
-    if (volumeVisibility && !this->ArrayForColoring)
+    this->CycleScalars(vtkF3DRendererWithColoring::CycleType::ARRAY_INDEX);
+    hasColoring =
+      this->Importer->GetInfoForColoring(this->UseCellColoring, this->ArrayIndexForColoring, info);
+  }
+
+  if (hasColoring && !this->ColorTransferFunctionConfigured)
+  {
+    this->ConfigureRangeAndCTFForColoring(info);
+    this->ColorTransferFunctionConfigured = true;
+  }
+
+  // Handle surface geometry
+  bool geometriesVisible = this->UseRaytracing || (!this->UseVolume && !this->UsePointSprites);
+  const auto& actorsAndMappers = this->Importer->GetGeometryActorsAndMappers();
+  for (size_t i = 0; i < actorsAndMappers.size(); i++)
```

Comment:
Let's try to use structured bindings here:
```cpp
for (auto& [actor, mapper] : actorsAndMappers)
```

Reply:
you'll lose access to `i` though

Reply:
and i need access to i.

Reply:
can't want for C++20 :P